gptel.el

```diff
@@ -912,6 +912,60 @@ Later plists in the sequence take precedence over earlier ones."
         (setq rtn (plist-put rtn p v))))
     rtn))
 
+;; https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)
+(defun gptel--uuid ()
+  "Generate a UUID v4-1."
+  (format "%04x%04x-%04x-4%03x-8%03x-%04x%04x%04x"
+          (random #x10000) (random #x10000)
+          (random #x10000)
+          (random #x1000)
+          (random #x1000)
+          (random #x10000) (random #x10000) (random #x10000)))
+
+(defun gptel--machine-id ()
+  "Generate a machine ID."
+  (let ((hex-chars "0123456789abcdef")
+        (length 65)
+        hex)
+    (dotimes (_ length)
+      (setq hex (nconc hex (list (aref hex-chars (random 16))))))
+    (apply #'string hex)))
+
+(cl-defun gptel--url-retrieve (url &key method data headers)
+  "Retrieve URL synchronously with METHOD, DATA and HEADERS."
+  (let ((url-request-method (if (eq method'post) "POST" "GET"))
+        (url-request-data (encode-coding-string (gptel--json-encode data) 'utf-8))
+        (url-mime-accept-string "application/json")
+        (url-request-extra-headers
+         `(("content-type" . "application/json")
+           ("editor-plugin-version" . "CopilotChat.nvim/2.0.0")
+           ("user-agent" . "CopilotChat.nvim/2.0.0")
+           ("editor-version" . "Neovim/0.10.0")
+           ,@headers)))
+    (with-current-buffer (url-retrieve-synchronously url)
+      (goto-char url-http-end-of-headers)
+      (gptel--json-read))))
+
+(defun gptel--restore (file)
+  "Restore saved object from FILE."
+  (when (file-exists-p file)
+    (car (read-from-string
```

Comment:
I suggest switching to the simpler `(read (current-buffer))` instead of converting the buffer to a string and then running `read-from-string`.

If this is not desirable for some reason, please let me know.